lookout/style/format/uast_stability_checker.py

```diff
@@ -0,0 +1,253 @@
+                self._log.debug("Code generator can't generate code: %s", repr(e.args))
+                unsafe_preds.append(i)
+                continue
+            parent_after, errors_after = parse_uast(
+                stub, content_after, filename="", language=self._feature_extractor.language)
+            if errors_after:
+                unsafe_preds.append(i)
+                continue
+            if not self.check_uasts_are_equal(parent_before, parent_after):
+                if self._debug:
+                    self._log.debug(
+                        "Bad prediction\nfile:%s\nDiff:\n%s\n\n", vnode_y.path,
+                        "\n".join(line for line in difflib.unified_diff(
+                            file_content[start:end].splitlines(), content_after.splitlines(),
+                            fromfile="original", tofile="suggested", lineterm="")))
+                unsafe_preds.append(i)
+        self._log.info("%d filtered out of %d with changes", len(unsafe_preds), changes.shape[0])
+        unsafe_preds = frozenset(unsafe_preds)
+        safe_preds = numpy.array([i for i in range(len(y)) if i not in unsafe_preds])
+        vnodes_y = [vn for i, vn in enumerate(list(vnodes_y)) if i not in unsafe_preds]
+        return y[safe_preds], y_pred[safe_preds], vnodes_y, rule_winners[safe_preds], safe_preds
+
+    def check(
+            self, y: numpy.ndarray, y_pred: numpy.ndarray, vnodes_y: Sequence[VirtualNode],
+            vnodes: Sequence[VirtualNode], files: Union[Sequence[File], File],
+            stub: "bblfsh.aliases.ProtocolServiceStub",
+            vnode_parents: Mapping[int, bblfsh.Node], node_parents: Mapping[int, bblfsh.Node],
+            rule_winners: numpy.ndarray, grouped_quote_predictions: QuotedNodeTripleMapping,
+    ) -> _check_return_type:
+        """
+        Filter the model's predictions that modify the UAST apart from changing Node positions.
+
+        :param y: Numpy 1-dimensional array of labels.
+        :param y_pred: Numpy 1-dimensional array of predicted labels by the model.
+        :param vnodes_y: Sequence of the labeled `VirtualNode`-s corresponding to labeled samples.
+        :param vnodes: Sequence of all the `VirtualNode`-s corresponding to the input.
+        :param files: File or Sequence of File-s with content, uast and path.
+        :param stub: Babelfish GRPC service stub.
+        :param vnode_parents: `VirtualNode`-s' parents mapping as the LCA of the closest
+                               left and right babelfish nodes.
+        :param node_parents: Parents mapping of the input UASTs.
+        :param rule_winners: Numpy array of the index of the winning rule for each sample.
+        :param grouped_quote_predictions: Quotes predictions (handled differenlty from the rest).
+        :return: List of predictions indices that are considered valid i.e. that are not breaking
+                 the UAST.
+        """
+        if isinstance(files, File):
```

Comment:
I'm not against this optimization (the other way being to just apply the splitting code in the case of a single file and pay the small overhead for the sake of simplicity) but I'd rather keep a list in all cases and have a test for len == 1. This way we don't encounter the counter-intuitive `files=file` and we avoid union in types which keeps mypy useful.

Reply:
This is confusing, yes